README
NAME
    Catalyst::View::RRDGraph - RRD Graph View Class

SYNOPSIS
    use the helper to create your View myapp_create.pl view RRDGraph
    RRDGraph

    from the controller:

      sub routine :Local {
        my ($self, $c) = @_;
    
        $c->stash->{'graph'} = [
                "--lower-limit", "0",
                "--start", "end-1d",
                "--vertical-label", "My Label",
                "--height", 600,
                "--width", 300,
                "DEF:Data=/path/to/rrd.rrd:data:AVERAGE",
                "AREA:Data#0000FF:Data "
        ];
        $c->forward('MyApp::View::RRDGraph');
      }

DESCRIPTION
    This view generates RRD graph images from the graph defintion placed in
    the stash. The controller is responsable of placing an ARRAYREF in
    $c-stash->{'graph'}> with the same data as to generate a graph with the
    RRDs module, except for *filename*, that will be automatically generated
    by the view.

    It doesn't depend on RRDs, but *does* need it to work (and for the tests
    to pass). You can install RRDs, for instance, using Alien::RRDtool, or
    compiling it manually from <http://oss.oetiker.ch/rrdtool/>

CONFIGURATION
    Configurations for the view are:

  IMG_DIR
    Directory to generate temporary image files. Defaults to /tmp/

  IMG_FORMAT
    Image format for the generated files. 'PNG' by default.

  ON_ERROR_SERVE
    On error, if this config value is set, the file to which it points will
    be served (so you can serve an "error image" file to the user).
    Alternately, it can be set to a code reference, that will called with
    $self, $c and $error. You can then generate your own content in this
    handler. Default (leaving undefined) is to throw an expception.

    See <http://oss.oetiker.ch/rrdtool/doc/rrdgraph.en.html> for more info.

METHODS
  new
    Constructor.

  process
    Called internally by Catalyst when the view is used.

AUTHOR
        Jose Luis Martinez
        CPAN ID: JLMARTIN
        CAPSiDE
        jlmartinez@capside.com
        L<http://www.pplusdomain.net>

THANKS
    To Ton Voon for sending in patches, tests, and ideas.

    Alexander Kabenin

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

SEE ALSO
    RRDs: <http://oss.oetiker.ch/rrdtool/prog/RRDs.en.html>

    RRD graph docs: <http://oss.oetiker.ch/rrdtool/doc/rrdgraph.en.html>,
    <http://oss.oetiker.ch/rrdtool/doc/rrdgraph_data.en.html>,
    <http://oss.oetiker.ch/rrdtool/doc/rrdgraph_graph.en.html>



Changes
Revision history for Perl module Catalyst::View::RRDGraph

0.09      2014-02-11 23:23:16+01:00 Europe/Madrid
    - Use File::Spec to build paths in tests

0.08      2012-12-07 17:56:21 Europe/Madrid
    - Catalyst::Plugin::ConfigLoader no longer needed to test

0.07      2012-11-19 00:48:53 Europe/Madrid
    - Skip tests if RRDs not found
    - Minor cosmetic changes

0.06      2012-10-26 14:53:30 Europe/Madrid
    - Released with Dist::Zilla
    - RRDs not a prerequisite since it will make fail some installers
    - Documentation modified accordingly

0.05 Mon Feb 07 20:46:00 2011
    - Fix RT #63114. Got away without "use Catalyst::Exception" for some time. Not now.

0.04 Thu Mar 25 15:20:00 2010
    - use MRO::Compat (patch courtesy Ton Voon)
    - change to Module::Install so that build pre-reqs can be separated from runtime pre-reqs

0.03 Sun Oct 26 21:35:00 2008
    - use base instead of use parent in test suite

0.02 Sun Oct 19 14:10:10 2008
    - Throw Catalyst Exceptions instead of die
    - RRDs returns a 0 byte image file and doesn't set RRDs::error. Detect and throw Exception
    - new ON_ERROR_SERVE attribute

0.01 Sun Aug  3 14:23:16 2008
    - original version; created by ExtUtils::ModuleMaker 0.51


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.009.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "RRD Graph View Class",
  "AUTHOR" => "Jose Luis Martinez Torres <jlmartin\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catalyst-View-RRDGraph",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::View::RRDGraph",
  "PREREQ_PM" => {
    "Catalyst::Plugin::Static::Simple" => 0,
    "Catalyst::Runtime" => "5.7",
    "MRO::Compat" => 0
  },
  "TEST_REQUIRES" => {
    "Catalyst::Test" => 0,
    "Test::MockObject" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.09",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Catalyst::Plugin::Static::Simple" => 0,
  "Catalyst::Runtime" => "5.7",
  "Catalyst::Test" => 0,
  "MRO::Compat" => 0,
  "Test::MockObject" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Catalyst/View/RRDGraph.pm
package Catalyst::View::RRDGraph;
{
  $Catalyst::View::RRDGraph::VERSION = '0.09';
}

use strict;
use warnings;

use base 'Catalyst::View';

use RRDs;
use File::Temp qw();
use MRO::Compat;

use Catalyst::Exception;

sub new {
    my ($class, $c, $arguments) = @_;
    my $config = {
        'IMG_DIR' => '/tmp/',
        'IMG_FORMAT' => 'PNG',
        'ON_ERROR_SERVE' => undef,
        %{ $class->config },
        (defined($arguments)?%{$arguments}:()),
    };

    my $self = $class->next::method(
        $c, { %$config },
    );

    $self->config($config);

    return ($self);
}

sub process {
    my ($self, $c) = @_;

    my $props = $c->stash->{'graph'};
    die "No graph in the stash" if (not defined $props);
    die "graph must be an ARRAYREF" if (ref($props) ne 'ARRAY');

    my $tempfile = File::Temp->new( TEMPLATE => 'cat_view_rrd_XXXXXX',
                                    DIR => $self->config->{'IMG_DIR'},
                                    SUFFIX => '.' . lc($self->config->{'IMG_FORMAT'}));

    RRDs::graph($tempfile->filename,
                '--imgformat', $self->config->{'IMG_FORMAT'},
                @$props);

    if (RRDs::error) {
        $self->_handle_error($c, RRDs::error);
        return;
    }
    if (-s $tempfile->filename == 0) {
        $self->_handle_error($c, "RRDgraph is 0 bytes");
        return;
    }

    $c->serve_static_file($tempfile->filename);
}

sub _handle_error {
    my ($self, $c, $error) = @_;
    if (not defined $self->config->{'ON_ERROR_SERVE'}){
        Catalyst::Exception->throw($error);
    } elsif (ref($self->config->{'ON_ERROR_SERVE'}) eq 'CODE') {
        #Call the custom handler
        $self->config->{'ON_ERROR_SERVE'}($self, $c, $error);
    } else {
        $c->log->error($error);
        $c->serve_static_file($self->config->{'ON_ERROR_SERVE'});
    }

}

#################### main pod documentation begin ###################
## Below is the stub of documentation for your module. 
## You better edit it!


=head1 NAME

Catalyst::View::RRDGraph - RRD Graph View Class

=head1 SYNOPSIS

use the helper to create your View myapp_create.pl view RRDGraph RRDGraph

from the controller:

  sub routine :Local {
    my ($self, $c) = @_;

    $c->stash->{'graph'} = [
            "--lower-limit", "0",
            "--start", "end-1d",
            "--vertical-label", "My Label",
            "--height", 600,
            "--width", 300,
            "DEF:Data=/path/to/rrd.rrd:data:AVERAGE",
            "AREA:Data#0000FF:Data "
    ];
    $c->forward('MyApp::View::RRDGraph');
  }

=head1 DESCRIPTION

This view generates RRD graph images from the graph defintion placed in the
stash. The controller is responsable of placing an ARRAYREF in
B<$c->stash->{'graph'}> with the same data as to generate a graph with the RRDs
module, except for I<filename>, that will be automatically generated by the
view.

It doesn't depend on RRDs, but I<does> need it to work (and for the tests to
pass). You can install RRDs, for instance, using L<Alien::RRDtool>, or
compiling it manually from L<http://oss.oetiker.ch/rrdtool/>

=head1 CONFIGURATION

Configurations for the view are:

=head2 IMG_DIR

Directory to generate temporary image files. Defaults to B</tmp/>

=head2 IMG_FORMAT

Image format for the generated files. 'PNG' by default. 

=head2 ON_ERROR_SERVE

On error, if this config value is set, the file to which it points will be
served (so you can serve an "error image" file to the user). Alternately, it
can be set to a code reference, that will called with B<$self>, B<$c> and
B<$error>. You can then generate your own content in this handler. Default
(leaving undefined) is to throw an expception.

See L<http://oss.oetiker.ch/rrdtool/doc/rrdgraph.en.html> for more info.

=head1 METHODS

=head2 new

Constructor.

=head2 process

Called internally by Catalyst when the view is used.

=head1 AUTHOR

    Jose Luis Martinez
    CPAN ID: JLMARTIN
    CAPSiDE
    jlmartinez@capside.com
    L<http://www.pplusdomain.net>

=head1 THANKS

To Ton Voon for sending in patches, tests, and ideas.

Alexander Kabenin

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=head1 SEE ALSO

RRDs: L<http://oss.oetiker.ch/rrdtool/prog/RRDs.en.html>

RRD graph docs:
L<http://oss.oetiker.ch/rrdtool/doc/rrdgraph.en.html>,
L<http://oss.oetiker.ch/rrdtool/doc/rrdgraph_data.en.html>,
L<http://oss.oetiker.ch/rrdtool/doc/rrdgraph_graph.en.html>

=cut

#################### main pod documentation end ###################


1;
# The preceding line will help the module return a true value

